library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java

```
@@ -0,0 +1,511 @@
+      {0, '#', '$', '§', 'Ä', 'Ö', 'Ü', '^', '_', '°', 'ä', 'ö', 'ü', 'ß'},
+      /* Swedish/Finnish/Hungarian (100%) */
+      {0, '#', '¤', 'É', 'Ä', 'Ö', 'Å', 'Ü', '_', 'é', 'ä', 'ö', 'å', 'ü'},
+      /* Italian (100%) */
+      {0, '£', '$', 'é', '°', 'ç', '»', '¬', '#', 'ù', 'à', 'ò', 'è', 'ì'},
+      /* French (100%) */
+      {0, 'é', 'ï', 'à', 'ë', 'ê', 'ù', 'î', '#', 'è', 'â', 'ô', 'û', 'ç'},
+      /* Portuguese/Spanish (100%) */
+      {0, 'ç', '$', '¡', 'á', 'é', 'í', 'ó', 'ú', '¿', 'ü', 'ñ', 'è', 'à'},
+      /* Czech/Slovak (60%) */
+      {0, '#', 'u', 'c', 't', 'z', 'ý', 'í', 'r', 'é', 'á', 'e', 'ú', 's'},
+      /* reserved (English mapping) */
+      {0, '£', '$', '@', '«', '½', '»', '¬', '#', '­', '¼', '¦', '¾', '÷'},
+
+      /* for latin-2 font */
+      /* Polish (100%) */
+      {0, '#', 'ñ', '±', '¯', '¦', '£', 'æ', 'ó', 'ê', '¿', '¶', '³', '¼'},
+      /* German (100%) */
+      {0, '#', '$', '§', 'Ä', 'Ö', 'Ü', '^', '_', '°', 'ä', 'ö', 'ü', 'ß'},
+      /* Estonian (100%) */
+      {0, '#', 'õ', '©', 'Ä', 'Ö', '®', 'Ü', 'Õ', '¹', 'ä', 'ö', '¾', 'ü'},
+      /* Lettish/Lithuanian (90%) */
+      {0, '#', '$', '©', 'ë', 'ê', '®', 'è', 'ü', '¹', '±', 'u', '¾', 'i'},
+      /* French (90%) */
+      {0, 'é', 'i', 'a', 'ë', 'ì', 'u', 'î', '#', 'e', 'â', 'ô', 'u', 'ç'},
+      /* Serbian/Croation/Slovenian (100%) */
+      {0, '#', 'Ë', 'È', 'Æ', '®', 'Ð', '©', 'ë', 'è', 'æ', '®', 'ð', '¹'},
+      /* Czech/Slovak (100%) */
+      {0, '#', 'ù', 'è', '»', '¾', 'ý', 'í', 'ø', 'é', 'á', 'ì', 'ú', '¹'},
+      /* Rumanian (95%) */
+      {0, '#', '¢', 'Þ', 'Â', 'ª', 'Ã', 'Î', 'i', 'þ', 'â', 'º', 'ã', 'î'},
+  };
+
+  private static final int[] CHARSET_INDEXES = {0, 4, 2, 6, 1, 5, 3, 7};
+
+  private static void checkParity(byte[] data, int offset, int len) {
+    for (int i = 0; i < len; ++i) {
+      int a = data[offset + i] & 0xFF;
+      if ((PARITY_TABLE[a] & 32) == 0) {
+        Log.w(TAG, "parity error at " + (offset + i));
```

shouldn't decoding fail here rather than just logging and continuing?

Same below for unrecoverable hamming codes